File: charuconstruction/frame.py
import cv2
import numpy as np


class Frame:
    def __init__(self, frame: np.ndarray):
        self._frame = frame
        self._grayscale_frame = self._ensure_grayscale()

        self._video_writer = None

    def frame_from(self, new_frame: "Frame"):
        """Overwrite the current frame with a new one."""
        self._frame = new_frame._frame.copy()
        self._grayscale_frame = new_frame._grayscale_frame.copy()

    def get(self, grayscale: bool = False) -> np.ndarray:
        """
        Get the underlying image array.

        Parameters:
            grayscale (bool): Whether to return the grayscale version of the frame.
        Returns:
            np.ndarray: The image array.
        """
        if grayscale:
            return self._grayscale_frame
        return self._frame

    def show(
        self,
        max_width=1280,
        max_height=720,
        grayscale: bool = False,
        wait_time: int = None,
    ) -> bool:
        """
        Show the frame in a window, resized to fit within max dimensions.

        Parameters:
            max_width (int): Maximum width of the displayed window.
            max_height (int): Maximum height of the displayed window.
            grayscale (bool): Whether to show the grayscale version of the frame.
            wait_time (int): Time in milliseconds to wait for a key event. If None, waits indefinitely.
        Returns:
            bool: True if the window is visible, False if it is closed.
        """
        to_show = self._grayscale_frame if grayscale else self._frame

        h, w = to_show.shape[:2]

        # Compute scaling factor to fit max size
        scale_w = max_width / w
        scale_h = max_height / h
        scale = min(scale_w, scale_h, 1.0)  # never upscale

        if scale < 1.0:
            new_w = int(w * scale)
            new_h = int(h * scale)
            resized = cv2.resize(to_show, (new_w, new_h))
        else:
            resized = to_show

        cv2.imshow("frame", resized)
        cv2.waitKey(wait_time)

        try:
            is_visible = (
                cv2.getWindowProperty("frame", cv2.WND_PROP_VISIBLE) >= 1
            )
            return is_visible

        except cv2.error:
            # Window was closed
            return False

    def _ensure_grayscale(self) -> np.ndarray:
        """Ensure the frame is in grayscale format."""
        if len(self._frame.shape) == 2:
            return self._frame
        elif len(self._frame.shape) == 3:
            return cv2.cvtColor(self._frame, cv2.COLOR_BGR2GRAY)
        else:
            raise ValueError("Invalid image shape")

    def start_recording(self, output_path: str, fps: int = 30):
        """Start recording the frame to a video file."""
        if self._video_writer is not None:
            raise RuntimeError("Recording already in progress")

        fourcc = cv2.VideoWriter_fourcc(*"mp4v")
        h, w = self._frame.shape[:2]
        self._video_writer = cv2.VideoWriter(output_path, fourcc, fps, (w, h))

    def add_frame_to_recording(self):
        """Add the current frame to the recording."""
        if self._video_writer is None:
            raise RuntimeError("Recording not started")
        self._video_writer.write(self._frame)

    def stop_recording(self):
        """Stop recording and release the video writer."""
        if self._video_writer is None:
            raise RuntimeError("Recording not started")
        self._video_writer.release()
        self._video_writer = None

File: charuconstruction_flutter/lib/modules/charucos/frame.dart
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:dartcv4/dartcv.dart';

class Frame {
  final Mat _image;
  Mat? _grayscaleImage;
  VideoWriter? _videoWriter;

  Frame(Mat image) : _image = image;

  ///
  /// Get the image as a Mat object.
  /// [grayscale] Whether to return the grayscale version of the image.
  /// If true, the grayscale version will be cached for future calls.
  ///
  Mat get({bool grayscale = false}) {
    return grayscale
        ? (_grayscaleImage ??= _convertToGrayscale(_image))
        : _image;
  }

  ///
  /// Dispose of the frame's resources. This should be called when the frame is no longer needed to free up memory.
  ///
  void dispose() {
    _image.dispose();
    _grayscaleImage?.dispose();
  }

  ///
  /// Convert the frame to a displayable format and show it in a window.
  /// The window will be automatically resized to fit the frame while respecting the specified maximum dimensions.
  /// [maxWidth] Maximum width of the displayed window.
  /// [maxHeight] Maximum height of the displayed window.
  /// [grayscale] Whether to show the grayscale version of the frame.
  ///
  /// Returns the frame as a byte array in PNG format, which can be used for
  /// further processing, showing or saving to disk. If the conversion fails, null is returned.
  ///
  Uint8List? toBytes({
    int maxWidth = 800,
    int maxHeight = 600,
    bool grayscale = false,
  }) {
    final frame = get(grayscale: grayscale);

    final height = frame.height;
    final width = frame.width;

    // Compute scaling factor to fit max size
    final scaleW = maxWidth / width;
    final scaleH = maxHeight / height;
    final scale = math.min(scaleW, scaleH); // never upscale

    late final Mat resized;
    if (scale < 1.0) {
      final newWidth = (width * scale).toInt();
      final newHeight = (height * scale).toInt();
      resized = resize(frame, (newWidth, newHeight));
    } else {
      resized = frame;
    }

    final (isSuccess, bytes) = imencode(".png", resized);
    return isSuccess ? bytes : null;
  }

  ///
  /// Close the window displaying the frame.
  ///
  void closeWindow() {
    destroyWindow('frame');
  }

  ///
  /// Start recording a video from the frame. The video will be saved to the specified
  /// filename with the given frames per second (fps).
  /// [filename] The name of the video file to save.
  /// [fps] The frames per second for the video.
  /// Throws an exception if a recording is already in progress.
  ///
  void startRecording(String filename, int fps) {
    if (_videoWriter != null) {
      throw Exception('Recording already in progress');
    }

    final frameSize = (_image.height, _image.width);
    _videoWriter = VideoWriter.fromFile(
      filename,
      'mp4v',
      fps.toDouble(),
      frameSize,
    );
  }

  ///
  /// Add the current frame to the ongoing video recording.
  /// Throws an exception if no recording is in progress.
  ///
  void addFrameToRecording() {
    if (_videoWriter == null) throw Exception('No recording in progress');

    _videoWriter!.write(_image);
  }

  ///
  /// Stop the ongoing video recording and save the file. After calling this method, a new recording can be started.
  /// Throws an exception if no recording is in progress.
  ///
  void stopRecording() {
    if (_videoWriter == null) throw Exception('No recording in progress');

    _videoWriter?.release();
    _videoWriter = null;
  }
}

///
/// Ensure the frame is in grayscale format.
///
Mat _convertToGrayscale(Mat image) {
  // TODO Check how to test for the number of channels in the image
  if (image.channels == 1) {
    return image;
  } else if (image.channels == 3) {
    return cvtColor(image, COLOR_BGR2GRAY);
  } else {
    throw ArgumentError('Unsupported number of channels: ${image.channels}');
  }
}
